### ISSUE TITLE
[Bug?]: Prisma built-in dataloader not working for GraphQL resolvers

### ISSUE THREAD

**Issue by jotalanusse** [redwoodjs/redwood#11736]
### What's not working?

Recently, while optimizing our API service, we encountered the N+1 problem in our GraphQL resolvers. This issue occurs when a separate database query is executed for each related record, resulting in multiple queries being made to retrieve additional information for associated data.

Upon further investigation, we referred to the Prisma documentation, which provides guidance on addressing this issue ([link](https://www.prisma.io/docs/orm/prisma-client/queries/query-optimization-performance#solving-the-n1-problem)). It suggests leveraging Prisma features, such as the fluent API, to enable the database to automatically batch requests and reduce the number of queries.

The issue is that, despite following the documentation and implementing these optimizations, the queries to the database are still not being batched as expected.

We are starting to suspect that the way Redwood, or its underlying GraphQL engine, executes resolvers might be interfering with the built-in data loaders provided by Prisma.

### How do we reproduce the bug?

GraphQL schema definition:
```gql
type Truck {
  id: String!
  name: String!
  organizationId: String!
  organization: Organization!
  routeRuns: [RouteRun]!
  createdAt: DateTime!
}
```

Main resolver:
```ts
export const trucksByOrganization: QueryResolvers['trucksByOrganization'] =
  async ({ organizationId }) => {
    return db.truck.findMany({ where: { organizationId } });
  };
```

Field resolvers:
```ts
export const Truck: TruckRelationResolvers = {
  organization: (_args, { root }) => db.truck.findUnique({ where: { id: root.id } }).organization(),
  routeRuns: (_args, { root }) => db.truck.findUnique({ where: { id: root.id } }).routeRuns(),
};

```

In the code above, the same databse instance is shared for all resolvers, where:
```ts
const db = new PrismaClient();
```

Given the setup described above (using `findUnique` and the fluent API), automatic batching should function as expected, but for some reason, it does not. Due to the private nature of the project, I cannot share the code directly. However, if a reproduction is needed to test this issue, I am willing to create one and provide it.

### What's your environment? (If it applies)

```shell
System:
  OS: Linux 6.8 Ubuntu 22.04.5 LTS 22.04.5 LTS (Jammy Jellyfish)
  Shell: 5.1.16 - /bin/bash
Binaries:
  Node: 18.20.4 - /tmp/yarn--1732636433916-0.04046345284337405/node
  Yarn: 1.22.19 - /tmp/yarn--1732636433916-0.04046345284337405/yarn
Browsers:
  Chrome: 131.0.6778.85
npmPackages:
  @redwoodjs/api-server: 6.6.4 => 6.6.4 
  @redwoodjs/auth-firebase-setup: 6.6.4 => 6.6.4 
  @redwoodjs/cli: 6.6.4 => 6.6.4 
  @redwoodjs/core: 6.6.4 => 6.6.4 
  @redwoodjs/internal: 6.6.4 => 6.6.4
```

### Are you interested in working on this?

- [ ] I'm interested in working on this

**Comment by mschmidt01** [comment]
@jotalanusse,

While investigating a similar problem, we discovered something interesting: there might be a discrepancy between what logs suggest and what's actually happening at the database level.

In our case, Prisma middleware logs showed multiple separate `findUnique` calls, suggesting an N+1 problem:
```
Building.findUnique took 0.006 seconds – {"where":{"id":"00000001-1011-0003-0091-000000000000"},...}
Building.findUnique took 0.006 seconds – {"where":{"id":"00000001-1012-0003-0099-000000000000"},...}
```

However, when we checked the raw database logs, we found Prisma was actually batching these into a single query:
```
SELECT "public"."Building"."id", ...  
FROM "public"."Building"  
WHERE "public"."Building"."id" IN ($1, $2, $3, ..., $36)  
```

Could you check your raw database logs to see if something similar is happening in your case? It might just be a logging issue rather than a problem with the actual dataloader implementation.

**Comment by awinograd** [comment]
@mschmidt01 I'm seeing this issue as well and the database logs to show duplicate queries being executed.  Here's my situation:


Resolvers
```typescript
export const book: QueryResolvers['book'] = ({ id }) => {
  return db.book.findUnique({
    select: { id: true },
    where: { id },
  })
}

export const Book: BookRelationResolvers = {
  pages: (_obj, { root }) => {
    return db.book
      .findUnique({ select: { id: true }, where: { id: root.id } })
      .pages()
  },
}
```

Query
```graphql
query Redwood {
  book(id: "cmdf6p1ro0000axkp0csy2zzz") {
    id
    pages {
      id
    }
  }
}
```


Database logs
```
2025-07-22 23:50:32.178 UTC [87] LOG:  execute s0: SELECT "public"."Book"."id" FROM "public"."Book" WHERE ("public"."Book"."id" = $1 AND 1=1) LIMIT $2 OFFSET $3
2025-07-22 23:50:32.178 UTC [87] DETAIL:  Parameters: $1 = 'cmdf6p1ro0000axkp0csy2zzz', $2 = '1', $3 = '0'
2025-07-22 23:50:32.180 UTC [87] LOG:  execute s0: SELECT "public"."Book"."id" FROM "public"."Book" WHERE ("public"."Book"."id" = $1 AND 1=1) LIMIT $2 OFFSET $3
2025-07-22 23:50:32.180 UTC [87] DETAIL:  Parameters: $1 = 'cmdf6p1ro0000axkp0csy2zzz', $2 = '1', $3 = '0'
2025-07-22 23:50:32.181 UTC [87] LOG:  execute s1: SELECT "public"."Page"."id", "public"."Page"."text", "public"."Page"."bookId", "public"."Page"."createdAt", "public"."Page"."updatedAt" FROM "public"."Page" WHERE "public"."Page"."bookId" IN ($1) OFFSET $2
2025-07-22 23:50:32.181 UTC [87] DETAIL:  Parameters: $1 = 'cmdf6p1ro0000axkp0csy2zzz', $2 = '0'
h
```

Based on the prisma docs, I wouldn't expect the book lookup to be duplicated.


Related, I dont think @jotalanusse 's example in the OP is supposed to work with the fluent API because of the `findMany` usage in the `trucksByOrganization` resolver

**Comment by awinograd** [comment]
So, it looks like if I go one level deeper, the prisma data loader / fluent API batching kicks in.  Query looks something like

```graphql
query Redwood {
  book(id: "cmdf6p1ro0000axkp0csy2zzz") {
    id
    pages {
      id
      paragraphs {
         id
      }
    }
  }
}
```

All of the resolvers at the `Book` level result in dup queries, but the pages resolvers are grouped.

ie if i'm selecting `x` fields from book and `n` pages are returned, i'm seeing `x+1+1` queries rather than `x+n+1`.

 So, I guess it's not catastrophic, but still a little odd the optimization doesn't kick in for the root level object resolvers 🤔

**Comment by Tobbe** [comment]
@jotalanusse I recommend you include `organization` and `routeRuns` in your initial prisma query if you always need to get those relation, rather than relying on the field resolvers

```ts
export const trucksByOrganization: QueryResolvers['trucksByOrganization'] =
  async ({ organizationId }) => {
    return db.truck.findMany({
      where: { organizationId },
      include: {
        organization: true,
        routeRuns: true,
      }
    })
  }
```
(I hope I got the syntax right. I didn't test this)

**Comment by Tobbe** [comment]
@awinograd And same advice applies to you too :) 

```ts
export const book: QueryResolvers['book'] = ({ id }) => {
  return db.book.findUnique({
    select: { id: true },
    where: { id },
    include: { pages: true },
  })
}
```

Something like the above if you need the pages for all books to be included.

And if you sometimes need the pages and sometimes don't, I'd create separate resolvers for the two variants. Like `export const book: ...` and `export const bookWithPages: ...`

**Comment by Tobbe** [comment]
> So, If i'm just doing query Books { books { id } } with your solution I have to load all of my pages

Yeah, that's why I suggested you create separate resolvers to suit your different use cases. But you're totally right in that there's no free lunch with that approach. 

I'm more than happy to learn about other approaches if you find something better! And/or if someone else has further input.

**Comment by ladderschool** [comment]
> [@Tobbe](https://github.com/Tobbe) and [@ladderschool](https://github.com/ladderschool) I appreciate your thoughts on this!
> 
> [@ladderschool](https://github.com/ladderschool) I've used `dataloader` extensively in a previous project and do like it a lot. I was hoping to get some of the n+1 fixes for "free" from the redwood setup and avoid having to pass them around everywhere in service tests. But good to know that for more mature apps, that's probably going to be where I end up.

I made a draft to show how I've implemented it, the app is a simple Books vs. Magazines comparison and to show how dataloaders work. I'll probably add performance tracking and debugging logs but it works.
